macro/derive/src/lib.rs

```diff
@@ -211,3 +211,45 @@ fn impl_from_variant(ast: &syn::DeriveInput) -> TokenStream {
     };
     gen.into()
 }
+
+/// [`VariantCount`] derives a function `const fn variant_count() -> usize` for enums
+/// that returns the count of variants in enum. E.g.
+/// ```
+/// #[derive(VariantCount)]
+/// enum MyEnum {
+///   First,
+///   Second(i32),
+///   Third {
+///     some: str,
+///     stuff: usize
+///   }
+/// }
+///
+/// assert_eq!(MyEnum::variant_count(), 3)
+/// ```
+///
+/// # Panics
+/// When derive attribute target is not an enum
+//
+// TODO: remove when https://github.com/rust-lang/rust/issues/73662
+// or alternative stabilizes
+#[proc_macro_derive(VariantCount)]
+pub fn variant_count_derive(input: TokenStream) -> TokenStream {
```

Comment:
I recommend using an existing crate: https://crates.io/crates/variant_count

Reply:
Actually `vairant_count` doesn't generate docs for derived constant and I don't think there's a way to suppress a lint in derived code without making it module-level.

Reply:
Anyway I like the idea of `associated constant` instead of `const fn`